pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -1309,3 +1338,89 @@ func initializeAnnotations(cd *hivev1.ClusterDeployment) {
 		cd.Annotations = map[string]string{}
 	}
 }
+
+// mergePullSecrets takes pull secretes in to account and returns the merged pull request
+func (r *ReconcileClusterDeployment) mergePullSecrets(cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) (string, error) {
+
+	// For code readability let's call the pull secret in cluster deployment config as local pull secret
+	localPullSecret, _ := controllerutils.LoadSecretData(r.Client, cd.Spec.PullSecret.Name, cd.Namespace, corev1.DockerConfigJsonKey)
+
+	// Check if global pull secret from env as it comes from hive config
+	globalPullSecret := os.Getenv("GLOBAL_PULL_SECRET")
+
+	switch {
+	case globalPullSecret != "" && localPullSecret != "":
+		// Merge local pullSecret and globalPullSecret. If both pull secrets have same registry name
+		// then the merged pull secret will have registry secret from local pull secret
+		pullSecret, err := controllerutils.MergeJsons(globalPullSecret, localPullSecret, cdLog)
+		if err != nil {
+			errMsg := "unable to merge global pull secret with local pull secret"
+			cdLog.WithError(err).Error(errMsg)
+			return "", fmt.Errorf("%s : %v", errMsg, err)
+		}
+		return pullSecret, nil
+	case globalPullSecret != "":
+		return globalPullSecret, nil
+	case localPullSecret != "":
+		return localPullSecret, nil
+	default:
+		errMsg := "clusterdeployment must specify pull secret since hiveconfig does not specify a global pull secret"
+		cdLog.Error(errMsg)
+		return "", fmt.Errorf("%s", errMsg)
+	}
+}
+
+// updatePullSecretInfo adds pull secret information in cluster deployment and cluster deployment status
+func (r *ReconcileClusterDeployment) updatePullSecretInfo(pullSecret string, cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) error {
+	var secretName string
+	var noExistingPullSecretObj bool
+
+	existingPullSecretObj := &corev1.Secret{}
+
+	if cd.Status.PullSecret.Name != "" {
+		secretName = cd.Status.PullSecret.Name
+		err := r.Get(context.TODO(), types.NamespacedName{Name: secretName, Namespace: cd.Namespace}, existingPullSecretObj)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				noExistingPullSecretObj = true
+			} else {
+				cdLog.WithError(err).Error("Error getting pull secret from cluster deployment")
+				return err
+			}
+		}
+	} else {
+		secretName = apihelpers.GetResourceName(cd.Name, "pull-secret-"+utilrand.String(8))
```

Comment:
Do not apply a random string to the end of the name. Let the api server do that for you by using `generateName` instead of `name`.

Reply:
Do we know the generated name immediately after creating it or do we need to figure it out with more queries?

Reply:
The object returned from the create call will have the generated name.

https://github.com/kubernetes/client-go/blob/4ed54556a14abf94c7aaabe2bd77fe4c591c9331/kubernetes/typed/core/v1/secret.go#L110

Reply:
I am still not sure how to use `generateName`. Can you point me to some sample code?